### ISSUE TITLE
Unable to catch the data-feed drop event.

### ISSUE THREAD

**Issue by cryptoeraser** [cryptoeraser/simple-websocket-externalData#2]
With the current setup, `ws.js` is not able to detect when the data stream stops. Since the `emission()` function is bound to the `file_watcher`, the moment the `data_writer.js` is killed and the stream to the file stops, the emission event is taken out.

Under such circumstances, I was not able to implement a way that will alert the user about an interrupted data stream. Ideally, the goal is to catch this event on the server side and update the corresponding flag `feed_active` inside the transmission container object to `false`: 
```
...
'records': {
        'client_input': false,
        'server_update': false,
        'feed_active': true,
...
```

A possible solution that I have thought of is to use a UUID number and cache the previous data. In that way the `prev_UUID` and the `current_UUID` maybe compared. In cases where the `UUID` values are the same, we might  flag this as an interrupted data feed. But I do not think this is a very elegant approach :)

Any ideas?

**Comment by diegoperini** [comment]
This solution is actually source specific but assuming it can be abstracted away, you have basically 3 options.

1. For first and the easiest, move the data writer operation to the same process with the reader. Then, notify the client when the connection is dropped via the error callback your API provider calls.

or

2. Establish a second file in which data writer first deletes its content and then writes a timestamp say each second. Reader process should check this file every X (X = your timeout threshold) seconds and read the timestamp. If it is above the threshold, connection is dropped. `setInverval` will do. This concept is called **heartbeat** and you can read about it online.

or

3. Use the builtin [cluster](https://nodejs.org/api/cluster.html) to create your processes. Your master process can keep track of the created child processes (status, whether they died) and pass messages between them. A simple heartbeat would do the trick here too. This is the most difficult and hardcore solution but the amount of plumbing here is relatively big.

---

Since you are using your own dummy code to simulate a source, you have to also simulate the connection drop in the method 1 which is the same problem you are trying to solve, "recursion"! Use this method for non dummy, real life data sources. 

For the dummy data source, use method 2. After you are done with your first fully featured app (months later), start reading the docs given in method 3 the day you are going to deploy the app on a cloud server

**Comment by cryptoeraser** [comment]
Thanks. I am going to try this inside a new dev branch. [dev-1.1.2.0](https://github.com/cryptoeraser/simple-websocket-externalData/tree/dev-1.1.2.0)

**Comment by cryptoeraser** [comment]
Hello,

I think I have solved the problem! :) Not sure which of the above I have used though. I definitely used the `setInterval()` approach, so probably number `2` of the solutions.

Here is a Vimeo [demo](https://vimeo.com/299116959) of how it is working now.

So, even with the data stream down, the client keeps receiving some form of `void` data and also the flags indicating the `data_feed_down` state. Once the feed is back, normal data continues to flow, but this time with an updated flag indicating a `data_feed_ok` state.

I realized (based on the answer [here](https://stackoverflow.com/questions/53137126/fs-watch-events-and-websockets-catching-no-change-situations/53152951?noredirect=1#comment93205714_53152951)) that since the `file.watcher` is forcing the websocket block to evaluate constantly on file change, I needed a similar mechanics, something that will keep cooking the websocket block but this time with no data. So I added a `data_feed_loop` and also a global `DATA_FEED_IS_ACTIVE` flag.

This might be a slightly dirty solution, but seems to be working. Thanks for all the help @diegoperini :)
